[IMP] account: discount allocation

hupo-odoo opened the pull request:
Provides the possibility for the user to enter defaults accounts for discounts allocation. If provided, discounts on Customer Invoices and Vendor BIlls will be allocation on the specified account instead of the product account.

Example:
Create a journal entry with one product with the following attributes: quantity: 1
price unit: 100
discount 5

We expect the following account move line:
Without an account set in the settings for discount allocation:
ACCOUNT                                             DEBIT     CREDIT
700000 Sales rendered in Belgium (marchandises)      0         95
400000 Trade debtors within one year - Customer      95        0

With discount allocation account set to "657000 Discount Given":
ACCOUNT                                             DEBIT     CREDIT
700000 Sales rendered in Belgium (marchandises)      0         95
700000 Sales rendered in Belgium (marchandises)      0         5
657000 Discounts Given                               5         0
400000 Trade debtors within one year - Customer      95        0

Note: the discount is added on another line than the one of the product even though it is the same account because it does not include taxes and analytic distributions.

task: 3278827




---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/133286).

william-andre commented on addons/account/models/account_move.py:
```diff
@@ -1983,6 +1983,13 @@ def _synchronize_business_models(self, changed_fields):
     # DYNAMIC LINES
     # -------------------------------------------------------------------------
 
+    def get_discount_allocation_account(self):
```
Doesn't need to be public
```suggestion
    def _get_discount_allocation_account(self):
```

william-andre replied:
I wouldn't put it in `# DYNAMIC LINES`. `# HELPER METHODS` would be better IMO

william-andre commented on addons/account/models/res_config_settings.py:
```diff
@@ -191,6 +191,22 @@ class ResConfigSettings(models.TransientModel):
+        string='Vendor Bills Discounts Account',
+        readonly=False,
+        related='company_id.account_discount_income_allocation_id',
+        domain="[('account_type', 'in', ('income', 'expense'))]",
```
Why expense?

chklop replied:
I can't speak for every Accounting practice in the world, but I've seen people using negative line in Sales to reduce amounts (for example in depreciations or cash discounts). I think we should not restrict the choice too much here.

william-andre replied:
So you would have a line on an income account, and then you would do a "negative" discount on an expense account? :exploding_head:
How would that work regarding the total sum D-C anyways?

william-andre replied:
But I think I understand what you mean, ok for keeping it then

william-andre commented on addons/account/models/account_move_line.py:
```diff
@@ -960,6 +964,56 @@ def _compute_all_tax(self):
+                frozendict({
+                    'move_id': line.move_id.id,
+                    'account_id': line.account_id.id,
+                    'display_type': 'discount',
```
The keys of the fronzendict should be the same as the ones from `discount_allocation_key`

I know it isn't like that on some other sync functions (like for EPD), and that is a mistake (causing useless re-computations and consuming SQL ids)

hupo-odoo replied:
Thanks for the explanation. Indeed I wasn't sure if the both keys were supposed to contain the same fields. I moved the `display_type: discount` to the dict below, but I realized it also work without it (the created lines have the display_type set to `discount` without the need to specify it. Should I keep it there for explicitly indicating that it is this value we want or should I just remove it ?

william-andre commented on addons/account/models/account_move_line.py:
```diff
@@ -960,6 +964,56 @@ def _compute_all_tax(self):
+            if not discount_allocation_account or line.display_type != 'product' or line.currency_id.is_zero(line.discount):
+                continue
+
+            discounted_amount_currency = line.quantity * line.price_unit * line.discount/100
```
Why not multiplying by `line.move_id.direction_sign` here already?

william-andre reviewed (commented):
Mostly details, good job going through this harder part of the code :+1:

chklop commented on addons/account/views/res_config_settings_views.xml:
```diff
@@ -273,6 +273,19 @@
                                     </div>
                                 </setting>
+                                <setting string="Separate discount accounts on invoices"
+                                         title="If empty, the discount will be discounted directly on the invoice account. If set, discount on invoices will be realized in separate accounts.">
```
Agreed, it's a more precise description.

hupo-odoo commented:
Thank you @william-andre for the review. I've made the requested changes and also added a small test.

william-andre reviewed (approved):
Good work!

We should just wait on https://github.com/odoo/odoo/pull/133267 and rebase to adapt the new fields the same way

hupo-odoo commented:
Hello @william-andre I have rebased on the above mentionned PR. I have set the new field `_needed` and `_key` as not exportable and added them in the `check_field_access_rights` function. I hope I did what was expected
